### PR TITLE
API Implement composite react fields / PopoverField

### PR DESCRIPTION
**Pull request by tractorcow**
This change is broken into two main changes:

 - Cleanup composite fields, removing lots of legacy code, improve php documentation and general code quality and housekeeping.
 - Add new PopoverField that will produce a nice popup in react; This is useful for hiding advanced form actions behind a composite holder.
 
 
image

Should be merged with https://github.com/silverstripe/silverstripe-cms/pull/1540 to fix tabs in the CMS.

### CODE REVIEW

**Comment by dhensby** [comment]
Hmmm. What's with the test failures?

**Comment by dhensby** [comment]
This is what's up with them: https://github.com/npm/npm/issues/13284

**Comment by dhensby** [comment]
Lots of codeclimate issues :/

**Comment by tractorcow** [comment]
I've fixed up a bunch of codeclimate issues.

CSS issues... I'd need @clarkepaul to look at since I didn't do that part of this pull request.

**Comment by dhensby** [comment]
It needs a rebase now :(

**Comment by tractorcow** [comment]
Rebased.

**Comment by dhensby** [comment]
@clarkepaul happy?

**Comment by clarkepaul** [comment]
Will just take a look

**Comment by dhensby** [comment]
@tractorcow codeclimate is not happy :/

**Comment by tractorcow** [comment]
@clarkepaul please help... it's not liking your CSS :P

**Comment by clarkepaul** [comment]
I've already checked it from a design and frontend perspective and its all good from me. From the code (JS) side I can't peer review. 

FYI The test popup has been removed from this branch, but the functionality is there still to add one back which makes it a bit harder to test.

File: admin/client/src/components/FormAction/FormAction.js [hunk 1]
```diff
@@ -9,26 +9,33 @@ class FormAction extends SilverStripeComponent {
...
 
   /**
+   * Get props for the button
+   *
+   * @returns {Object}
+   */
+  getButtonProps() {
+    // Merge attributes
+    return Object.assign({},
+      typeof this.props.attributes === 'undefined' ? {} : this.props.attributes,
```
Comment (flamerohr) [inline admin/client/src/components/FormAction/FormAction.js]: I'd suggestion adding a `propTypes = { attributes:  React.PropTypes.object, }` and `defaultProps = { attributes: {}, }`
So you get cleaner autonomous code in React

File: admin/client/src/components/FormAction/FormAction.js [hunk 2]
```diff
@@ -113,10 +161,4 @@ FormAction.propTypes = {
   extraClass: React.PropTypes.string,
 };
 
-FormAction.defaultProps = {
```
Comment (flamerohr) [inline admin/client/src/components/FormAction/FormAction.js]: how come this was removed?

File: admin/client/src/components/PopoverField/PopoverField.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    const overlay = (
+      <Popover id={`${this.props.id}_Popover`} className={`fade in popover-${placement}`}
+        title={this.getPopoverTitle()}
+      >
+        {this.props.children}
+      </Popover>
+    );
+    // If no text is specified, use ... with xl icon style
+    const buttonClasses = ['btn', 'btn-secondary', 'btn--no-focus'];
+    if (!this.props.title) {
```
Comment (flamerohr) [inline admin/client/src/components/PopoverField/PopoverField.js]: wouldn't this use `this.getPopoverTitle()` as well? (one `this.props.title` below too)

**Comment by clarkepaul** [comment]
I'm fixing a few linting issues

File: admin/client/src/components/FormAction/FormAction.js (hunk 2 above)
Comment (tractorcow) [inline admin/client/src/components/FormAction/FormAction.js]: - type is now passed in through attributes nested
- bootstrapButtonStyle doesn't have a set default anymore; It depends on the name to determine fallback (save action will have a different style to other actions).
- disabled; I'll put this back though. ;P

File: admin/client/src/components/PopoverField/PopoverField.js (hunk 3 above)
Comment (tractorcow) [inline admin/client/src/components/PopoverField/PopoverField.js]: Title is title of the button, popover title is the title of the popup window. You can have one and not the other, so it shouldn't inherit from either.

File: admin/client/src/components/FormAction/FormAction.js (hunk 1 above)
Comment (tractorcow) [inline admin/client/src/components/FormAction/FormAction.js]: +1

**Comment by dhensby** [comment]
1 bit of feedback then it's ready

**Comment by tractorcow** [comment]
@dhensby you didn't leave any feedback. :)

**Comment by tractorcow** [comment]
Please don't merge just yet... @clarkepaul is fixing CSS issues.

**Comment by tractorcow** [comment]
CSS issues fixed. @sminnee this is now ready for review on passing. :)

**Pull request merged**